docs/spec_draft.md

```diff
@@ -348,6 +349,93 @@ logical operation.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.batch_norm_training
+
+### Semantics
+
+Computes mean and variance across all dimensions except for the `feature_index`
+dimension and normalizes the `operand` tensor producing `output`, `batch_mean`
+and `batch_var` tensors. More formally, this operation can be expressed as a
+decomposition to existing StableHLO operations using Python-like syntax as
+follows:
+
+```python
+def compute_mean(operand, feature_index):
+  aggregate = reduce(inputs = operand, init_values = 0, body = lambda x,y : x+y,
+               dimensions = [ i for i in range(rank(operand)) if i != feature_index ])
+  return divide(aggregate, size(operand) / dim(operand, feature_index))
+
+def compute_variance(operand, feature_index):
+  centered_operand = subtract(operand, compute_mean(operand, feature_index))
+  centered_operand_squared = mul(centered_operand, centered_operand)
+  return compute_mean(centered_operand_squared, feature_index)
+
+def batch_norm_training(operand, scale, offset, epsilon, feature_index):
+  # Compute `mean` and `variance` (as 1-dimensional tensors)
+  mean = compute_mean(operand, feature_index)
+  variance = compute_variance(operand, feature_index)
+
+  # Perform normalization with computed `mean` and `variance` instead of
+  # depending on `mean` and `variance` provided in the input like
+  # `batch_norm_inference`
+  return batch_norm_inference(operand, scale, offset, mean,
+                              variance, epsilon, feature_index)
+```
```

Comment:
Proposed refactoring of the code:

```Python
def compute_mean(operand, feature_index):
  (sum,) = reduce(
      inputs=[operand],
      init_values=[0],
      dimensions=[i for i in range(rank(operand)) if i != feature_index],
      body=lambda x, y: add(x, y))
  divisor = constant(num_elements(operand) / dim(operand, feature_index))
  divisor_bcast = broadcast_in_dim(mean, [], shape(sum))
  return divide(sum, divisor_bcast)

def compute_variance(operand, feature_index):
  mean = compute_mean(operand, feature_index)
  mean_bcast = broadcast_in_dim(mean, [feature_index], shape(operand))
  centered_operand = subtract(operand, mean_bcast)
  return compute_mean(mul(centered_operand, centered_operand), feature_index)

def batch_norm_training(operand, scale, offset, epsilon, feature_index):
  mean = compute_mean(operand, feature_index)
  variance = compute_variance(operand, feature_index)
  return batch_norm_inference(operand, scale, offset, mean, variance, epsilon,
                              feature_index)
```

Feedback:
  1. For `reduce`, follow the order of inputs in the ODS.
  2. Respect varargs for `inputs`, `init_values` and `results` of `reduce`.
  3. Change `x + y` to `add(x, y)`.
  4. Inline `centered_operand_squared` to avoid introducing a new name.
  5. Remove comments in `batch_norm_training` - they don't seem to provide additional context, unlike with `batch_norm_inference`.
  6. Add broadcasts where needed.
  7. Reformat using `pyformat`.

Reply:
1. I had seed the order, but I wanted to follow the order in which they appear in MLIR syntax, which is why I used named arguments (will change)
2. I had noticed that as well but I thought single values are treated as array of 1 element
3. I had thought of that as well but I thought people would confuse it with stablehlo add, come to think about it that we want stablehlo add
4. sure
5. sure
6. I thought division by scalar is automatically broadcasted
7. sure

Thanks for the feedback!

Reply:
1a. I think that named arguments are a nice touch because they help understand what each argument means - that's less obvious without looking into the `reduce` spec than e.g. with `add` or `mul`.
1b/2. Let's use the ordering and the types from the spec - I'd say that MLIR implementation details are just implementation details as far as the spec goes.
6. Generally, StableHLO doesn't allow implicit broadcasting. The only two instances that support them are described in #442, and I think I'd like to get rid of them in the future.

Reply:
thanks